src/cutlass_mlp.cu

```diff
@@ -314,6 +327,606 @@ void CutlassMLP<T>::backward_impl(
 	}
 }
 
+// ======================= backward_backward_input_impl =======================
+// compute 2nd order dact
+template <typename T>
+__global__ void compute_activation_backward_backward(uint32_t n_elements, Activation activation, T* p, T* res) {
+	uint32_t i = blockIdx.x * blockDim.x + threadIdx.x;
+	if (i >= n_elements) return;
+
+	switch (activation) {
+		case Activation::Softplus:
+			float K_ACT = 10.0;
+			float tmp = (float)p[i] * K_ACT;//
+			if (tmp > 10.0) {
+				tmp = 10.0;
+			} else if (tmp < -15.0) {
+				tmp = -15.0;
+			}
+
+			float exp_tmp = expf(tmp);
+			float pow_tmp = (exp_tmp + 1.0) * (exp_tmp + 1.0);
+			float ddoutputdp_dp = exp_tmp / pow_tmp * K_ACT;
+			res[i] = (T)ddoutputdp_dp;
+			return;
+
+		case Activation::ReLU:
+			res[i] = 0.0;
+			return;
+
+		default:
+			// ERROR: this activation currently is not supported
+			res[i] = 0.0;
+			return;
+	}
+
+	return;
+}
+
+// activation Softplus 2nd order derivative in 1D: doutputdp_2
+template <typename T>
+__global__ void compute_ddoutputdp_dp_dLdoutput(uint32_t n_elements, T* dL_doutput, T* ddoutputdp_dp, T* doutputdp_2) {
+	uint32_t i = blockIdx.x * blockDim.x + threadIdx.x;
+	if (i >= n_elements) return;
+
+	doutputdp_2[i] = ddoutputdp_dp[i] * dL_doutput[i];
+
+	return;
+}
+
+// fuse the process of computing ddoutputdp_dp_2
+template <typename T>
+__global__ void fuse_ddoutputdp_dp(uint32_t n_elements, T* dL_doutput, T* w_x_dL2_ddL1dinput, T* ddoutputdp_dp, T* result) {
+	uint32_t i = blockIdx.x * blockDim.x + threadIdx.x;
+	if (i >= n_elements) return;
+	
+	T dL1doutput_x_w_x_dL2_ddL1dinput = dL_doutput[i] * w_x_dL2_ddL1dinput[i];
+	result[i] = ddoutputdp_dp[i] * dL1doutput_x_w_x_dL2_ddL1dinput;
+
+	return;
+}
+
+// element-wise add back to dL_dinput
+template <typename T>
+__global__ void element_wise_add(uint32_t n_elements, T* tmp, T* res) {
+	uint32_t i = blockIdx.x * blockDim.x + threadIdx.x;
+	if (i >= n_elements) return;
+
+	res[i] += tmp[i];
+}
+
+// element-wise copy from an CM tmp to RM dL_dinput
+// row: the tmp rows
+// col: the tmp cols
+template <typename T>
+__global__ void element_wise_copy_CM_RM(uint32_t n_elements, uint32_t row, uint32_t col, T* tmp, T* res) {
+	uint32_t i = blockIdx.x * blockDim.x + threadIdx.x;
+	if (i >= n_elements) return;
+
+	uint32_t idx_i = i % row;
+	uint32_t idx_j = i / row;
+	uint32_t i_RM = idx_i * col + idx_j;
+	res[i_RM] = tmp[i];
+}
+
+// compute d_doutputdp2_dinput
+template <typename T>
+__global__ void multiply_w_RM(uint32_t n_elements, uint32_t w_row, uint32_t w_col, T* weights, T* doutputdp_2, T* result) {
+	uint32_t i = blockIdx.x * blockDim.x + threadIdx.x;
+	if (i >= n_elements) return;
+
+	uint32_t idx_doutputdp = i % w_row;
+	T tmp = doutputdp_2[idx_doutputdp];
+	uint32_t i_RM = (i % w_row) * w_col + i / w_row;
+	result[i] = weights[i_RM] * tmp;
+}
```

Comment:
These should use `MatrixView` instead of raw pointers into matrices. This'll turn the kernels effectively into one-liners that you can implement inline below using `parallel_for_gpu_soa` / `parallel_for_gpu_aos` (depending on which access pattern is more performant). I recommend checking out `grid.h` as an example for how to use all these tools.